content/non-blocking-communication-pt1.rst

```diff
@@ -1,33 +1,324 @@
+
+This workflow will always work. This approach generalizes to larger
+rings, so long as alternating processes either send or respectively
+receive first. It also generalizes to multiple dimensions. However, it
+still requires synchronization between send and receive ranks. That
+synchronization is problematic when the time to send messages is
+comparable with either the amounts of compute work involved, or the
+variation in those amounts.
+
+Performant non-blocking MPI workflows for stencil applications
+--------------------------------------------------------------
+
+.. figure:: img/non-blocking-stylestencilworkflow.svg
+   :align: center
+
+   A correct non-blocking-style parallel stencil-computation
+   workflow. The MPI implementation may be able to do the
+   communication overlapped with the computation.
+
+This will also always work. It may be more efficient than the
+ring-style approach, but that depends on the implementation and
+environment, as well as the size of messages and amounts of compute
+that can participate in the overlapped region.
+
+The user is responsible for not using the memory area until after the
+message request has been waited upon.
+
+Non-blocking MPI send calls
+---------------------------
+
+An |term-MPI_Isend| is creates a send request and returns a request
+object. It may or may not have sent the message, or buffered it. The
+caller is responsible for not changing the buffer until after waiting
+upon the resulting request object.
+
+Call signature:
+
+.. code-block:: c
+
+    int MPI_Isend(const void* buf, int count, MPI_Datatype datatype,
+                  int dest, int tag, MPI_Comm comm, MPI_Request *request)
+
+.. note::
+
+    ``buf``, ``count`` and ``datatype`` describe the buffer to be sent
+    to ``dest`` rank of ``comm`` using tag ``tag``. The ``request`` object
+    that is returned must be used to wait on the communication later.
+
+Other calls exist for other sending modes familiar to you from
+point-to-point messages, including buffered, synchronous, and
+ready-mode sends. They are listed in the table below, along with
+links for more information.
+
+.. table:: Point-to-point communication functions
+   :widths: auto
+
+   +------------------------+------------------------+------------------+-------------------+
+   |                                                 |             Communication            |
+   +                                                 +------------------+-------------------+
+   |                                                 | Blocking         | Nonblocking       |
+   +========================+========================+==================+===================+
+   |                        | **Standard**           | |term-MPI_Send|  | |term-MPI_Isend|  |
+   +                        +------------------------+------------------+-------------------+
+   | **Mode**               | **Synchronous**        | |term-MPI_Ssend| | |term-MPI_Issend| |
```

Comment:
|term-MPI_Issend|, |term-MPI_Rsend|, |term-MPI_Irsend|, |term-MPI_Bsend|, and |term-MPI_Ibsend| are not defined. I think these can be simply monospace, no need to have them in the glossary.

Reply:
Also reasonable